eclipse-theia/theia
# plugin-ext: Change `PluginsKeyValueStorage` to not continuously access the disk

#### What it does

Fixes #8384

Added the use of a semaphore to lock access to the global plugin storage file, to avoid it being accessed in a disorderly fashion, causing potential corruption.

As per the issue discussion, it's possible that there is also a potential race condition caused by node's write function, not flushing its buffers to disk, but in this specific case, it was obvious that parallel calls to set() and get() were the main cause of the problems.

#### How to test

One way is to add ENTER/EXIT debug traces [1] to `packages/plugin-ext/src/main/node/plugins-key-value-storage.ts`, in `set()` and `get()` and then rebuild , run the browser tests ('yarn browser test`), and confirm that the accesses are now matched. i.e. "ENTER get()" should always be followed by "EXIT get()", and "ENTER set()" always followed by "EXIT set(). 

\[1\]
```  diff
diff --git a/packages/plugin-ext/src/main/node/plugins-key-value-storage.ts b/packages/plugin-ext/src/main/node/plugins-key-value-storage.ts
index 7200ce2e049..c0a7e91a325 100644
--- a/packages/plugin-ext/src/main/node/plugins-key-value-storage.ts
+++ b/packages/plugin-ext/src/main/node/plugins-key-value-storage.ts
@@ -57,6 +57,7 @@ export class PluginsKeyValueStorage {
     }
 
     async set(key: string, value: KeysToAnyValues, kind: PluginStorageKind): Promise<boolean> {
+        console.log('ENTER set()');
         try {
             await lock.acquire();
             const dataPath = await this.getDataPath(kind);
@@ -77,10 +78,12 @@ export class PluginsKeyValueStorage {
             return true;
         } finally {
             lock.release();
+            console.log('EXIT set()');
         };
     }
 
     async get(key: string, kind: PluginStorageKind): Promise<KeysToAnyValues> {
+        console.log('ENTER get()');
         try {
             await lock.acquire();
             const dataPath = await this.getDataPath(kind);
@@ -91,10 +94,12 @@ export class PluginsKeyValueStorage {
             return data[key];
         } finally {
             lock.release();
+            console.log('EXIT get()');
         };
     }
 
     async getAll(kind: PluginStorageKind): Promise<KeysToKeysToAnyValue> {
+        console.log('ENTER getAll()');
         try {
             await lock.acquire();
             const dataPath = await this.getDataPath(kind);
@@ -104,6 +109,7 @@ export class PluginsKeyValueStorage {
             return this.readFromFile(dataPath);
         } finally {
             lock.release();
+            console.log('EXIT getAll()');
         }
     }

```

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
